### ISSUE TITLE
unreachable_patterns false negative with or_patterns

### ISSUE THREAD

**Issue by tmiasko** [rust-lang/rust#76836]
The match remains exhaustive with highlighted pattern removed, but it does not produce any warnings:

```rust
#![feature(or_patterns)]

pub struct A;

pub enum B {
    X,
    Y,
}

pub enum E {
    A(A),
    B(B),
}

pub fn f(e: E) -> () {
    match e {
        E::B(B::X) => (),
        E::B(B::Y | B::X)
        //          ~~~~ 
        | E::A(A) => (),
    }
}
```

**Comment by Nadrieril** [comment]
I know where the problem comes from: this bit of code tries to handle unreachable sub-or-patterns diagnostics, but it's a bit of a hack. What remains is to come up with a better solution.
https://github.com/rust-lang/rust/blob/9832374f6e378971e1a933362cf9781b121bb845/compiler/rustc_mir_build/src/thir/pattern/_match.rs#L2041

@rustbot claim
